### PR TITLE
[dynamo] Lazy disable_dynamo API out-of-dynamo 

### PR DESCRIPTION
**Pull request by anijain2305**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #104397
* __->__ #104317
* #104276

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78

### CODE REVIEW

**Review by wconstab — COMMENTED** [review]
i would hesitate on merging this without having a discussion on the import plan.  I think both @jansel and @voznesenskym were hitting different circular import issues this week and maybe we should do something more about it than just locally solving the disable problem

**Comment by anijain2305** [comment]
@jansel and @voznesenskym Can you comment on what circular import issues you saw lately? And if this PR is relevant to it?

Note that this is different from @malfet 's PR - https://github.com/pytorch/pytorch/pull/104368 which lazily imports _dynamo when needed.

The problem this PR is trying to solve is limited to PyTorch codebase. Whenever, we want to put `disable` and `disallow_in_graph` inside core, there is a circular import problem.

File: torch/_compile.py
Comment (jansel) [inline torch/_compile.py:33]: return functools.partial(_disable_dynamo, recursive=recursive)

**Review by jansel — APPROVED** [review]

File: torch/_compile.py
Comment (anijain2305) [reply torch/_compile.py:33]: Ah, so clean!

**Comment by jansel** [comment]
> @jansel and @voznesenskym Can you comment on what circular import issues you saw lately? And if this PR is relevant to it?

My circular issue was in C++ symbols and unrelated to this.

File: torch/_compile.py
Comment (jansel) [inline torch/_compile.py:26]: I believe `disable` sets some properties on the return method.  Might also make sense to set them here.

**Review by wconstab — APPROVED** [review]
wasn't someone asking if we could import dynamo from torch?  would that solution make this simpler or not matter? i am ok with landing- we could continue to improve the import situation later regardless, and this would be good to unblock

**Comment by anijain2305** [comment]
> wasn't someone asking if we could import dynamo from torch? would that solution make this simpler or not matter? i am ok with landing- we could continue to improve the import situation later regardless, and this would be good to unblock

Importing dynamo from torch would not help with the problem we are trying to solve (atleast in context of this https://github.com/pytorch/pytorch/pull/104368). Even after that PR is merged, we would have circular imports issue while trying to disable functions inside core. This PR is trying to solve that problem.

**Comment by mlazos** [comment]
Thank you!! <3

**Review by mlazos — APPROVED** [review]

**Comment by anijain2305** [comment]
@pytorchhbot merge

**Pull request closed without merging**
